### ISSUE TITLE
libc: implement `getc()`

### ISSUE THREAD

**Issue by ycsin** [zephyrproject-rtos/zephyr#66943]
This was originally added as a ticket under POSIX, but `getc()` is not a POSIX function. It is from C89.

https://en.cppreference.com/w/c/io/fgetc

The expectation from the POSIX API is that this function is implemented as part of whatever C library is in use.

**Comment by cfriedt** [comment]
Is this not already in the minimal libc??

**Comment by cfriedt** [comment]
> I would like to take this. It would be helpful if you can guide and share some reference about this issue.

Hi @00thirdeye00 - `getc(FILE *stream)` shouldn't be too difficult.

From a very high level, it should be as simple as 
1. lock the fdtable entry associated with `stream` (and verify that it is valid)
2. read a character and report it back, or `EOF` if there is an error.
3. unlock the fdtable entry

Do we still have `zvfs_read()` or was that reverted? With that, it would be trivial to implement.

**Comment by 00thirdeye00** [comment]
@cfriedt hej, thank you, since i am new to zephyr, it can be helpful if you can provide any references to files, implementation and flow

**Comment by cfriedt** [comment]
@00thirdeye00 -

lib/os/fdtable.c <-- ftdable and some ZVFS functions
lib/c_lib/common/ <-- common libc components

https://docs.zephyrproject.org/latest/contribute/index.html

**Comment by seyoungjeong** [comment]
@00thirdeye00 @cfriedt has anybody started to work on this issue? Or can I take it?

**Comment by ycsin** [comment]
> [@00thirdeye00](https://github.com/00thirdeye00) [@cfriedt](https://github.com/cfriedt) has anybody started to work on this issue? Or can I take it?

Please feel free to do so.

- For a typical file, refer to @cfriedt comment above.
- For `stdin`, we need to somehow hook the console UART with the libc, but I'm quite confused with the implementation of `drivers/console/uart_console.c` - it looks kinda ad-hoc and probably needs to be detangled from the mcumgr handlers, the `subsys/console` might also need some rework. I was actually looking into that ytd (https://github.com/ycsin/zephyr/commit/ef0f8264261ba8672df2d346bbd24d1799df0737).

**Comment by seyoungjeong** [comment]
- Guess the `getc()` implementation would go under libc/minimal. And the definition would be `int getc(FILE* stream);`
- It sounds like that `FILE` handle could be possibly either file or UART handle?

**Comment by ycsin** [comment]
> * Guess the `getc()` implementation would go under libc/minimal.

Yes, if it is only used by minimal libc. If it's going to be shared by other libc then probably `lib/libc/common`

> And the definition would be `int getc(FILE* stream);`

Yes, you can read more about the definition here: https://en.cppreference.com/w/c/io/fgetc

> * It sounds like that `FILE` handle could be possibly either file or UART handle?

Anything that can be read from or written to can be abstracted as a 'FILE'

**Comment by seyoungjeong** [comment]
Sounds good.

**Comment by seyoungjeong** [comment]
I'm thinking to design and implement `getc()` using the same pattern as `putc()`. The design patten used for `putc()` leverages hooks(`_stdout_hook` for output) to abstract platform-specific behaviors. Similarly, `getc()` can use a `_stdin_hook` mechanism for input.

**Comment by stephanosio** [comment]
> This was originally added as a ticket under POSIX, but `getc()` is not a POSIX function. It is from C89.
> 
> https://en.cppreference.com/w/c/io/fgetc
> 
> The expectation from the POSIX API is that this function is implemented as part of whatever C library is in use.

Minimal libc, by definition, only exists to meet the needs of the Zephyr "kernel", as defined by the [Coding Guidelines Rule A.4](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-4-c-standard-library-usage-restrictions-in-zephyr-kernel).

Note that the Zephyr "kernel" is defined as follows in the Coding Guidelines Rule A.4:

> * Kernel (kernel)
> * OS Library (lib/os)
> * Architecture Port (arch)
> * Logging Subsystem (subsys/logging)

POSIX API is not part of the Zephyr "kernel" and, by design, the Zephyr POSIX subsystem was never meant to be and should never be used with the minimal libc.

@ycsin @cfriedt Given that Picolibc, the default standard C library for Zephyr, and all the other standard C libraries supported by Zephyr implement the `getc` function, I am in favour of closing this issue.

p.s. see also https://github.com/zephyrproject-rtos/zephyr/pull/93062#pullrequestreview-3018284678

**Comment by seyoungjeong** [comment]
As a PR creator I don't have any objection to close this issue. Good to know guide line for picolibc and minimal.

**Comment by cfriedt** [comment]
Works for me.

I think a lot of these issues dealing with libc-related features were unfortunately created by mistake.
